src/test/java/spoon/test/filters/FilterTest.java

```diff
@@ -640,82 +637,87 @@ public void testElementMapFunctionArray() throws Exception {
 	}
 	@Test
 	public void testElementMapFunctionNull() throws Exception {
+		// contract: when a function returns null, it is discarded at the next step
 		final Launcher launcher = new Launcher();
 		CtQueryImpl q = new CtQueryImpl().map((String s)->null);
-		List<String> list = q.setInput(null).list();
-		assertEquals(0, list.size());
-		
-		list = q.setInput("c").list();
+		List<String> list = q.setInput("c").list();
 		assertEquals(0, list.size());
 	}
 	@Test
 	public void testReuseOfQuery() throws Exception {
+		// contract: a query created from an existing element can be reused on other inputs
 		final Launcher launcher = new Launcher();
 		launcher.setArgs(new String[] {"--output-type", "nooutput","--level","info" });
 		launcher.addInputResource("./src/test/java/spoon/test/filters/testclasses");
 		launcher.run();
 		
-		class Context {
-			int count = 0;
-		}
-		Context context = new Context();
-
 		CtClass<?> cls = launcher.getFactory().Class().get(Tacos.class);
 		CtClass<?> cls2 = launcher.getFactory().Class().get(Tostada.class);
-		CtQuery q = cls.map((CtClass<?> c, CtConsumer<Object> out)->out.accept(c.getSimpleName()));
-		q.forEach((String name)->{
-			context.count++;
-			assertEquals(cls.getSimpleName(), name);
-		});
+
+		// by default the query starts with "cls" as input
+		CtQuery q = cls.map((CtClass c) -> c.getSimpleName());
+		// high-level assert
+		assertEquals(cls.getSimpleName(), q.list().get(0));
+		// low-level assert on implementation
 		assertEquals(1, ((CtQueryImpl)q).getInputs().size());
 		assertSame(cls, ((CtQueryImpl)q).getInputs().get(0));
-		//change the input of query to cls2
-		q.setInput(cls2).forEach((String name)->{
-			context.count++;
-			assertEquals(cls2.getSimpleName(), name);
-		});
+
+		// now changing the input of query to cls2
+		q.setInput(cls2);
 		//the input is still cls2
```

Comment:
I am evaluating the query twice intentionally. To assure the contract that call of setInput is persistent... that it is not reset after first call of list or forEach. So please add the second call of list again